stan/math/opencl/assign_events.hpp

```diff
@@ -86,6 +86,18 @@ inline void assign_event<write_buffer>(const matrix_cl* m,
   m->add_write_event(new_event);
 }
 
+// End of assign_events recursion
+template <typename T>
+inline void assign_events(const T& new_event) {}
+
+template <typename T, typename... Args>
+inline void assign_events(const cl::Event& new_event,
+   internal::to_const_matrix_cl_v<T>& m,
+   internal::to_const_matrix_cl_v<Args>&... args) {
+    assign_event<T>(m, new_event);
+    assign_events<Args...>(new_event, args...);
+}
```

Comment:
We should probably add some tests for this - this shouldn't even compile :P

Reply:
LoL which part? Is there some bad parameter pack expansion here?

Reply:
assign_event -> assign_events, also the order of the arguments are switched (I think you copied my comment, in fairness, which wasn't quite right).

Reply:
I wouldn't say "shouldn't compile" but would say, "is bad"

`assign_events` is the recursive part where `assign_event` is the actual assigning function

I can clean this up tonight to look correct

Reply:
Oh, I think they're supposed to have the same name. Should look like that example https://ideone.com/L4cqkA

Reply:
So I had an issue here and had to make a function as `assign_new_event` because when everything was `assign_event` the signature

```
inline void assign_event(const cl::Event& new_event,
                          internal::to_const_matrix_cl_v<T>& m,
                          internal::to_const_matrix_cl_v<Args>&... args)
```

Was the best template match instead of the function at the second to last part of the recursion. If you add

```cpp
// Turn on c++17 std
  constexpr std::size_t n = sizeof...(Args);
  printf("%lu", n);
```

to the recursive function and copy replace `assign_new_event` with `assign_event` you'll see the function goes on recursively 5ever. This happens because at the second to last recursion, template argument deduction chooses the recursive function over the function that ends the recursion.

Is there a simple way to go about this that I'm missing?